src/platform-aws.c

```diff
@@ -737,3 +737,105 @@ int platform_config_endpoint(struct fi_info *info, struct fid_ep* endpoint) {
 exit:
 	return ret;
 }
+
+static int get_rail_vf_idx(struct fi_info *info)
+{
+	char guid_file[256], guid[20];
+	char *ptr, *endptr;
+	int vf_idx;
+	FILE *fp;
+
+	snprintf(guid_file, sizeof(guid_file), "/sys/class/infiniband/%s/node_guid",
+		info->nic->device_attr->name);
+	fp = fopen(guid_file, "r");
+	if (fp == NULL) {
+		NCCL_OFI_WARN("Error opening file: %s", guid_file);
+		return -EIO;
+	}
+
+	if (fgets(guid, sizeof(guid), fp) == NULL) {
+		NCCL_OFI_WARN("Error reading file: %s", guid_file);
+		fclose(fp);
+		return -EIO;
+	}
+	fclose(fp);
+
+	/**
+	 * GUID is a 64-bit hex number with format:
+	 *
+	 * XXXX:XXXX:XXXX:XXXX
+	 *
+	 * The lowest 8 bits are the VF id.
+	 */
+
+	/* Find the final colon */
+	ptr = strrchr(guid, ':');
+	if (ptr == NULL) {
+		NCCL_OFI_WARN("Bad GUID format in %s", guid_file);
+		return -EINVAL;
+	}
+
+	/* ptr string now has format ":XXXX". Extract the final two digits as
+	   as the vf idx */
+	vf_idx = (int)strtol(ptr + 3, &endptr, 10);
```

Comment:
ok, now that I understand what you're doing, this needs some more sanity.  Does ptr + 3 actually point to something valid?  What if the format was XXX:XXXX:XXXX:X?

Reply:
Added appropriate checks

Reply:
You did not.  XXXX:XXXX:XXXXXXXX:X passes your checks, but will still explode.  Skip lines 770-772, and directly test the data you care about.

Reply:
Your example would be caught by the second part of this check: `ptr == NULL || ptr != &guid[14]`, since the last colon isn't in the right place.

Not sure what your ask is here. This code is relying on `strtol` to figure out whether the VF id at `ptr+3` is a valid integer, which seems like a reasonable way to check that (although I see now we need to update the check below for endptr). Are you asking that we check the entire GUID to make sure it matches the format we want?

Reply:
I fixed the check on endptr and changed the strlen check to be on ptr for clarity.

I think this is pretty robust. It doesn't verify the correct format of the entire GUID, but it should handle all error cases of the format of the last 8 bits correctly.

Reply:
I missed the 14 position check.  You're right it covers that, but now my complaint is that you've way overcomplicated the code.  If you're enfocing that the string be 18 characters long, character 14 is a colon, and everything that follows is something you can jam into strtol, then write the code that way and save a string parsing.

Reply:
Got it, makes sense. Updated.